presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxNAggregation.java

```diff
@@ -0,0 +1,175 @@
+        List<Type> inputTypes = ImmutableList.of(type, BIGINT);
+        MinMaxNStateSerializer stateSerializer = new MinMaxNStateSerializer(comparator, type);
+        Type intermediateType = stateSerializer.getSerializedType();
+        ArrayType finalType = new ArrayType(type);
+
+        List<ParameterMetadata> inputParameterMetadata =  ImmutableList.of(
+                new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, type),
+                new ParameterMetadata(INPUT_CHANNEL, BIGINT),
+                new ParameterMetadata(BLOCK_INDEX));
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(name, type, inputTypes),
+                inputParameterMetadata,
+                INPUT_FUNCTION.bindTo(comparator).bindTo(type),
+                null,
+                null,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION.bindTo(finalType),
+                MinMaxNState.class,
+                stateSerializer,
+                new MinMaxNStateFactory(),
+                finalType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(name, inputTypes, intermediateType, finalType, true, false, factory);
+    }
+
+    public static void input(BlockComparator comparator, Type type, MinMaxNState state, Block block, long n, int blockIndex)
+    {
+        if (block.isNull(blockIndex)) {
+            return;
+        }
+        TypedHeap heap = state.getTypedHeap();
+        if (heap == null) {
+            if (n <= 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "second argument of max_n/min_n must be positive");
+            }
+            heap = new TypedHeap(comparator, type, Ints.checkedCast(n));
+            state.setTypedHeap(heap);
+        }
+        long startSize = heap.getEstimatedSize();
+        heap.add(block, blockIndex);
+        state.addMemoryUsage(heap.getEstimatedSize() - startSize);
+    }
+
+    public static void combine(MinMaxNState state, MinMaxNState otherState)
+    {
+        TypedHeap otherHeap = otherState.getTypedHeap();
+        if (otherHeap == null) {
+            return;
+        }
+        TypedHeap heap = state.getTypedHeap();
+        if (heap == null) {
+            state.setTypedHeap(otherHeap);
+            return;
+        }
+        long startSize = heap.getEstimatedSize();
+        heap.addAll(otherHeap);
+        state.addMemoryUsage(heap.getEstimatedSize() - startSize);
+    }
+
+    public static void output(Type type, MinMaxNState state, BlockBuilder out)
+    {
+        TypedHeap heap = state.getTypedHeap();
+        if (heap == null || heap.isEmpty()) {
+            out.appendNull();
+            return;
+        }
+        BlockBuilder arrayBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus());
```

Comment:
I think this needs to use `type.getElementType().createBlockBuilder()` otherwise it's not going to work on array types